File: Binary Search/Java/Count the Number of Fair Pairs.java
// Problem link: https://leetcode.com/problems/count-the-number-of-fair-pairs/description
// Time: O(n log n)
// Space: O(n)
class Solution {
    public long countFairPairs(int[] nums, int lower, int upper) {
        long result = 0, largest, smallest;
        Arrays.sort(nums);

        for(int index = 0; index < nums.length; index++) {
            largest = this.binarySearch(nums, index, upper - nums[index] + 1);
            smallest = this.binarySearch(nums, index, lower - nums[index]);

            result += largest - smallest;
        }

        return result;
    }

    private int binarySearch(int[] nums, int index, int target) {
        int left = index + 1, right = nums.length - 1, mid;

        while(left <= right) {
            mid = (left + right) / 2;

            if(nums[mid] >= target) {
                right = mid - 1;
            }
            else {
                left = mid + 1;
            }
        }

        return right;
    }
}

File: Binary Search/Python/Count the Number of Fair Pairs.py
# Problem link: https://leetcode.com/problems/count-the-number-of-fair-pairs/description
# Time: O(n log n)
# Space: O(n)
class Solution:
    def countFairPairs(self, nums: List[int], lower: int, upper: int) -> int:
        nums.sort()
        result = 0

        for index in range(len(nums)):
            largest = self.binarySearch(nums, index, upper - nums[index] + 1)
            smallest = self.binarySearch(nums, index, lower - nums[index])

            result += largest - smallest
        
        return result

    def binarySearch(self, nums, index, target):
        left, right = index + 1, len(nums) - 1
        best = 0

        while left <= right:
            mid = (left + right) // 2

            if nums[mid] >= target:
                right = mid - 1
            else:
                left = mid + 1
            
        return right
